[lyuanww] iP

lyuanww opened the pull request:
**Duke**

Duke liberates your mind from the duty of memorizing your to-do list, offering the following advantages:

- Text-based
- Easy to learn
- ~~FAST~~ **SUPER FAST** to use

To get started, simply:

1. [Download it from here.](https://github.com/lyuanww/ip)
2. Add your tasks.
3. Let it manage your tasks for you 😉
   
And the best part? It's **FREE**!

### Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)


If you're a Java programmer, you can even use it to practice Java. Here's the main method:

```java
public static void main(String[] args) {
         Duke duke = new Duke("Duke", "src/main/java/resource/duke.txt");
         duke.run();
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

lyuanww committed:
Level-0

lyuanww committed:
Implement Level-1. Echo

lyuanww committed:
Implement Level-2. Add, List

lyuanww committed:
Implement Level-3. Mark as Done

lyuanww committed:
Implement Level-4. ToDos, Events, Deadlines

lyuanww committed:
Implement A-TextUiTesting: Automated Text UI Testing

lyuanww committed:
Implement Level-5: Handle Errors

lyuanww committed:
Implement Level-6: Delete

lyuanww committed:
Rename storage to tasklist

lyuanww committed:
Implement Level7.Save

lyuanww committed:
Merge branch 'branch-Level-7'

lyuanww committed:
Implement Level8.Dates and Times

lyuanww committed:
Merge branch 'branch-Level-8'

lyuanww committed:
Implement A-MoreOOP

lyuanww committed:
Merge branch 'add-more-oop'

lyuanww committed:
Implement A-Packages

lyuanww committed:
Merge branch 'add-gradle-support'

lyuanww committed:
Implement A-Gradle: Use Gradle

lyuanww committed:
Add checkstyle files

lyuanww committed:
Implement A-JUnit

lyuanww committed:
Change file not found message

lyuanww committed:
Implement JavaDoc

lyuanww committed:
Implement Follow the Coding Standard

lyuanww committed:
Implement Level-9: Find

lyuanww committed:
Merge branch 'branch-A-CodingStandard'

lyuanww committed:
Merge branch 'branch-Level-9'

lyuanww committed:
Resolve merge conflicts

LoMaply commented on src/main/java/duke/DateParser.java:
```diff
@@ -0,0 +1,38 @@
+    public static String convertDateTimeToString(LocalDateTime datetime) {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd-MM-yyyy HH:mm");
+        return datetime.format(formatter);
+    }
```
You could consider extracting out **formatter** as a static variable since it is used in both methods (so you only have to change one line if you ever decide to change formats)

Beyond that good job on extracting out the DateParser class

LoMaply commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,66 @@
+public abstract class Task {
+
+    private final String name;
+    private boolean done;
```
This boolelan field ought to be renamed "isDone" to be more clear.

LoMaply commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,66 @@
+     *
+     * @param display Whether to display a message indicating the task was marked as done.
+     */
+    public void markAsDone(boolean display) {
```
The variable name "display" for both this and unmarkAsDone might be a little unclear. Perhaps a more descriptive name such as "willDisplayMessage" although long, might be better for clarity?

LoMaply commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,158 @@
+     *
+     * @param taskDescription The description of the task.
+     * @param storage         The storage handler to save the task.
+     * @param ui              The user interface for displaying messages.
```
Generally your javadoc comments are well done, but I'm not sure if an extended space between the @param tag and it's accompanying text might mess something up

LoMaply commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,158 @@
+        }
+
+        System.out.println();
+        System.out.println("Here are the tasks in your list:");
```
Line 84 here is probably unnecessary. You can declare a new line using the \n character as follows:

`System.out.println("\nHere are the tasks in your list:");`

I noticed several other instances of an empty println statement that can be removed with this (Such as this file and the Ui class).

LoMaply commented on src/test/java/duke/StorageStub.java:
A javadocs comment explaining the use of this class in testing methods requiring a Storage object might help with clarity for why this fille exists.

LoMaply reviewed (commented):
Overall your code is quite well organised when it comes to classes. There were barely any coding standard violations. One general suggestion is to group classes into multiple packages, just to make it easier to find a specific class.

eg. Task, TaskList, ToDo, Event and Deadline can be under a "Task" package.

sopa301 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,100 @@
+ * by adding, marking as done, deleting, and listing tasks.
+ */
+public class Duke {
+    private final String name;
```
I like that you gave the bot a name! It might be useful when dealing with multiple bots.

sopa301 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,66 @@
+     */
+    public Task(String name) {
+        this.name = name;
+        this.done = false;
```
Perhaps a more boolean-ish name would work better like isDone?

sopa301 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,45 @@
+     *
+     * @return The formatted string for storage.
+     */
+    public String convertTaskToString() {
```
I'm not sure I like how convertTaskToString() and toString() sound really similar. Is there a way to differentiate them further, maybe like toDataString()?

sopa301 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,105 @@
+            }
+        }
+
+        if (fromIndex == -1 || toIndex == -1 || fromIndex == 0 || toIndex == 0 || fromIndex >= toIndex
```
I'm not sure I like the large number of conditions. Is there a way to condense them or abstract them so that the code would flow more naturally?

sopa301 reviewed (commented):
Overall, I like how the code is broken down into small and readable parts. However, grouping all classes into one big package might be difficult to manage as the project grows bigger. Consider having multiple sub-packages if it makes sense.

lyuanww committed:
Restructure files

lyuanww committed:
Refactor: add command classes

lyuanww committed:
Add comments to Command classes

lyuanww committed:
Modify comment for processInputIntoCommand

lyuanww committed:
Resolve PR comments

lyuanww committed:
Resolve PR comments - part 2

lyuanww committed:
Use CheckStyle

lyuanww committed:
Merge branch 'branch-A-CheckStyle'

lyuanww committed:
Fix checkstyle

lyuanww committed:
Implement Level-10: GUI

lyuanww committed:
Merge branch 'branch-Level-10'

lyuanww committed:
Fix Junit Test

lyuanww committed:
Add necessary changes to build JAR file

lyuanww committed:
Add approriate error handling during file creation

lyuanww committed:
Add assertions

Some files have pre-existing assumptions that should be verified during
runtime.

Assertions are added to Duke.java, Parser.java and TaskList.java. More
assertions are to be added if applicable.

Assertions are used to define assumptions about the program state so that the runtime can verify them.

lyuanww committed:
Merge pull request #1 from lyuanww/branch-A-Assertions

Add assertions

lyuanww committed:
Fix bug where message not displayed when task is marked as undone

Remove System.out.println in TaskList.java

lyuanww committed:
Implement A-CodeQuality

lyuanww committed:
Merge pull request #2 from lyuanww/branch-A-CodeQuality

Implement A-CodeQuality

lyuanww committed:
Fix checkstyle

lyuanww committed:
Implement Use Streams

lyuanww committed:
Merge pull request #3 from lyuanww/branch-A-Streams

Implement Use Streams

lyuanww committed:
Implement C-Sort

lyuanww committed:
Merge branch 'branch-C-Sort'

lyuanww committed:
Add styling to chatbot and add resizing property

lyuanww committed:
Merge branch 'branch-A-BetterGui'

lyuanww committed:
Implement: add personality

lyuanww committed:
Merge branch 'branch-A-Personality'

lyuanww committed:
Add Ui.png

lyuanww committed:
fix: bug where incorrect datetime causes null pointer exception

lyuanww committed:
fix: styling

lyuanww committed:
Update README.md

lyuanww committed:
change jar file name and fix minor styling

lyuanww committed:
fix: bug where program is not exited

lyuanww committed:
Remove hard-coded file path

lyuanww committed:
Add minor changes

lyuanww committed:
Fix styling

lyuanww committed:
Refine user guide

lyuanww committed:
Correct datetime format for ug
